### PR TITLE
Account for additional techniques for hiding content in `isVisible()`

### LINKED ISSUE #519: `isVisible()` does not account for elements pulled offscreen via `position`
`isVisible()` returns `true` for the following case:

```html
<style>
  .hidden {
    position: absolute;
    left: -9999px;
  }
</style>

<div class="hidden">This text is hidden</div>
```

#531 is needed for this.

- [ ] Follow up on DEV-11054.

### LINKED ISSUE #524: `isVisible()` does not account for text hidden via `text-indent`
`isVisible()` returns `true` for the following case:

```html
<style>
  .hidden {
    overflow: hidden;
    white-space: nowrap;
    text-indent: 100%;
  }
</style>

<div class="hidden">This text is hidden</div>
```

#532 is needed for this.

- [ ] Follow up on DEV-10828.
- [ ] Follow up on DEV-10887.

### PR DESCRIPTION
**Pull request by kasperisager**
Closes #519.
Closes #524.

### CODE REVIEW

File: packages/alfa-rules/src/common/predicate/is-offscreen.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          const { value: position } = style.computed("position");
+          const { value: left } = style.computed("left");
+          const { value: right } = style.computed("right");
+
+          if (position.value !== "static") {
+            for (const inset of [left, right]) {
+              switch (inset.type) {
+                case "length":
+                  return abs(inset.value) >= 9999;
+              }
```
Comment (Jym77) [inline packages/alfa-rules/src/common/predicate/is-offscreen.ts:34]: Why not looking also at `top` (I guess `bottom` is less doable due to infinity vertical scrolling)? (`position: absolute; top: -9999px` seems to be the preferred offscreen way in ACT rules at least).

File: packages/alfa-rules/src/common/predicate/is-offscreen.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+          const { value: position } = style.computed("position");
+          const { value: left } = style.computed("left");
+          const { value: right } = style.computed("right");
+
+          if (position.value !== "static") {
+            for (const inset of [left, right]) {
+              switch (inset.type) {
+                case "length":
+                  return abs(inset.value) >= 9999;
```
Comment (Jym77) [inline packages/alfa-rules/src/common/predicate/is-offscreen.ts:33]: Also, obvious prone to errors on huge displays, or with huge elements… Not sure we can do much better without a layout system.
(also, ACT rules is having `-999em` in at least two cases, which we won't catch here)

File: packages/alfa-rules/src/common/predicate/is-visible.ts [hunk 3]
```diff
@@ -6,107 +6,76 @@ import { Refinement } from "@siteimprove/alfa-refinement";
...
       and(
         isElement,
         and(not(isReplaced), (element) =>
           every(
-            element.children({ nested: true, flattened: true }),
+            element.children({
+              nested: true,
+              flattened: true,
+            }),
             not(isVisible(device, context))
```
Comment (Jym77) [inline packages/alfa-rules/src/common/predicate/is-visible.ts:75]: For the sake of readability of `isVisible`, I think we can go further in the subfunctions and extract these as two separate predicates.

File: packages/alfa-rules/test/common/predicate/is-visible.spec.tsx [hunk 4]
```diff
@@ -94,3 +94,229 @@ test("isVisible() returns true for replaced elements with no child", (t) => {
...
+  const text = h.text("");
+
+  t.equal(isVisible(text), false);
+});
+
+test("isVisible() returns false for a text node with no data", (t) => {
+  const text = h.text("");
+
+  t.equal(isVisible(text), false);
+});
```
Comment (Jym77) [inline packages/alfa-rules/test/common/predicate/is-visible.spec.tsx:172]: Duplicate test.
```suggestion
test("isVisible() returns false for a text node with no data", (t) => {
  const text = h.text("");

  t.equal(isVisible(text), false);
});
```

File: packages/alfa-rules/test/common/predicate/is-visible.spec.tsx [hunk 5]
```diff
@@ -94,3 +94,229 @@ test("isVisible() returns true for replaced elements with no child", (t) => {
...
+
+test("isVisible() returns false for a text node with only whitespace", (t) => {
+  const text = h.text(" ");
+
+  t.equal(isVisible(text), false);
+});
+
+test("isVisible() returns false for a text node with a transparent color", (t) => {
```
Comment (Jym77) [inline packages/alfa-rules/test/common/predicate/is-visible.spec.tsx:180]: For this and the following examples, I'm not sure the `<div>` element gets the same result as its text 🤔

**Review by Jym77 — CHANGES_REQUESTED** [review]

File: packages/alfa-rules/src/common/predicate/is-clipped.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+              whitespace.value === "nowrap"
+            ) {
+              const { value: indent } = style.computed("text-indent");
+
+              switch (indent.type) {
+                case "percentage":
+                  return abs(indent.value) >= 1;
+
+                case "length":
+                  return abs(indent.value) >= 999;
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-clipped.ts:63]: It's a heuristic for lack of a layout system. If we had a way of knowing the width of the element, we'd of course be able to account for that.

File: packages/alfa-rules/src/common/predicate/is-offscreen.ts (hunk 1 above)
Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-offscreen.ts:34]: As most pages grow downwards, not sideways, content should be offscreen'ed using `left` or `right` as the width of the viewport can be expected to remain constant. This article also talks about the perils of offscreen'ing using `top` and uses `left` in its examples: http://accessibilitytips.com/2008/03/04/positioning-content-offscreen/. This article from WebAIM uses `left` as well: https://webaim.org/techniques/css/invisiblecontent/

File: packages/alfa-rules/src/common/predicate/is-offscreen.ts (hunk 2 above)
Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-offscreen.ts:33]: These are computed `px` values so `em` values will have been resolved by the time this condition is reached. `-999em` can be expected to compute to at least `-9999px` for reasonable font sizes; all it takes is `10px`, after all.

File: packages/alfa-rules/test/common/predicate/is-visible.spec.tsx (hunk 5 above)
Comment (kasperisager) [reply packages/alfa-rules/test/common/predicate/is-visible.spec.tsx:174]: Very good point 👍 I've adjusted the test cases.

File: packages/alfa-rules/src/common/predicate/is-visible.ts (hunk 3 above)
Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-visible.ts]: Agreed, though I'll leave that to a separate task.

File: packages/alfa-rules/src/common/predicate/is-clipped.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+          const { value: height } = style.computed("height");
+          const { value: width } = style.computed("width");
+          const { value: x } = style.computed("overflow-x");
+          const { value: y } = style.computed("overflow-y");
+
+          if (
+            height.type === "length" &&
+            height.value <= 1 &&
+            width.type === "length" &&
+            height.value <= 1 &&
```
Comment (Jym77) [inline packages/alfa-rules/src/common/predicate/is-clipped.ts:35]:
```suggestion
            width.value <= 1 &&
```
🙈

File: packages/alfa-rules/src/common/predicate/is-clipped.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        if (isElement(node)) {
+          const style = Style.from(node, device, context);
+
+          const { value: height } = style.computed("height");
+          const { value: width } = style.computed("width");
+          const { value: x } = style.computed("overflow-x");
+          const { value: y } = style.computed("overflow-y");
+
+          if (
+            height.type === "length" &&
```
Comment (Jym77) [inline packages/alfa-rules/src/common/predicate/is-clipped.ts:32]: Also, wondering whether we should check for `height: 0%`. But maybe that's not really used for clipping.

**Review by Jym77 — CHANGES_REQUESTED** [review]

File: packages/alfa-rules/src/common/predicate/is-clipped.ts (hunk 7 above)
Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-clipped.ts:35]: Great catch! :facepalm:

File: packages/alfa-rules/src/common/predicate/is-clipped.ts (hunk 8 above)
Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-clipped.ts:32]: Good point!

**Review by Jym77 — APPROVED** [review]

**Pull request merged**